cpp/purify/h5reader.h

```diff
@@ -0,0 +1,237 @@
+  }
+
+#ifdef PURIFY_MPI
+  /// @brief Alternative constructor enabling MPI-collective behaviour
+  H5Handler(const std::string& filename, const sopt::mpi::Communicator& comm)
+      : _comm(&comm),
+        _fap(MPIFileAccess()),
+        _dtp(MPIDataTransfer()),
+        _file(filename, HighFive::File::ReadOnly, _fap) {
+    _datalen = _slicepos = _slicelen = _batchpos = 0;
+  }
+#endif
+
+  /// Method to read the entire dataset
+  template <typename T = double>
+  std::vector<T> read(const std::string& label) const {
+    auto dataset = _file.getDataSet(label);
+    return dataset.read<std::vector<T>>();
+  }
+
+  /// Method to read a dataset slice with
+  /// slices evenly split across MPI ranks
+  template <typename T = double>
+  std::vector<T> distread(const std::string& label) {
+    if (!_comm) throw std::runtime_error("No MPI-collective reading enabled!");
+    _loadDataSet(label);
+    std::vector<T> data;
+    data.reserve(_slicelen);
+    _ds[label].select({_slicepos}, {_slicelen}).read(data, _dtp);
+    return data;
+  }
+
+  /// Method to stochastically draw a subset
+  /// from the distributed dataset slice
+  template <typename T = double>
+  std::vector<T> stochread(const std::string& label, size_t batchsize, bool shuffle = false) {
+    if (!_comm) throw std::runtime_error("No MPI-collective reading enabled!");
+
+    _loadDataSet(label);
+    if (shuffle) _shuffle();
+
+    std::vector<T> data;
+    data.reserve(batchsize);
+    // account for wrap around near
+    // the edges of the slice
+    size_t pos = _batchpos;
+    while (batchsize) {
+      std::vector<T> tmp;
+      size_t len = std::min(batchsize, _slicepos + _slicelen - pos);
+      _ds[label].select({pos}, {len}).read(tmp, _dtp);
+      data.insert(data.end(), std::make_move_iterator(std::begin(tmp)),
+                  std::make_move_iterator(std::end(tmp)));
+      pos = _slicepos;
+      batchsize -= len;
+    }
+    return data;
+  }
+
+ private:
+  void _loadDataSet(const std::string& label) {
+    if (_ds.find(label) != _ds.end()) return;
+
+    _ds[label] = std::move(_file.getDataSet(label));
+    const auto& dims = _ds[label].getDimensions();
+    size_t len = dims.at(0);
+    if (len == 0) throw std::runtime_error("Dataset has zero length!");
+    if (len < _comm->size()) throw std::runtime_error("Not enough data for each MPI rank!");
+    if (_datalen == 0) {
+      _datalen = len;
+      // Determine starting position and length of the data slice
+      _slicelen = _datalen / _comm->size();
+      _slicepos = _comm->rank() * _slicelen;
+      if (_comm->rank() == _comm->size() - 1) {
+        _slicelen += _datalen % _comm->size();
+      }
+    } else if (len != _datalen) {
+      throw std::runtime_error("Inconsistent dataset length!");
+    }
+  }
+
+  void _shuffle() {
+    std::uniform_int_distribution<size_t> uni(_slicepos, _slicepos + _slicelen - 1);
+    _batchpos = uni(_rng);
+  }
+
+  const sopt::mpi::Communicator* _comm;
```

Comment:
This type is only defined if sopt is compiled with MPI turned on, so my compilation breaks if I have not compiled SOPT with MPI on even if I compiled purify with MPI off as well

Reply:
Ouff, good point. It looks like we currently have no cmake-level check that Sopt has MPI enabled if Purify enables MPI. Does the case Purify::MPI::ON with Sopt::MPI::OFF even make sense or should that be vetoed at config time?

Reply:
I suspect it doesn't make sense, since we seem to rely on the `sopt::mpi::communicator` for Purify's MPI business 🤔

Reply:
I'ma tweak the Purify CMake to throw an error if Purify is configured with MPI enabled but Sopt wasn't. Shout if that's offensive.